lib/connectors/mongodb/connector.rb

```diff
@@ -56,30 +70,166 @@ def initialize(configuration: {})
         @direct_connection = configuration.dig(:direct_connection, :value)
       end
 
-      def yield_documents
+      def yield_documents(job_description = {}, &on_doc_serialization)
+        advanced_filter_config = extract_filter_config(job_description)
+        rules = extract_rules(job_description)
+
+        check_filter_config!(rules, advanced_filter_config)
+
+        function, args = setup_db_function_with_args(advanced_filter_config)
+
+        call_db_function_on_collection(function, args, &on_doc_serialization)
+      end
+
+      private
+
+      def extract_rules(job_description)
+        Utility::Commons.return_if_present(job_description.dig(:filtering, :rules), DEFAULT_RULES)
+      end
+
+      def extract_filter_config(job_description)
+        Utility::Commons.return_if_present(job_description.dig(:filtering, :advanced_config), DEFAULT_ADVANCED_CONFIG)
+      end
+
+      def setup_db_function_with_args(filter_config)
+        # default parameters
+        function = FIND
+        args = NO_ARGS
+
+        function, args = setup_find(filter_config) if find_present?(filter_config)
+
+        function, args = setup_aggregate(filter_config) if aggregate_present?(filter_config)
+
+        [function, args]
+      end
+
+      def check_filter_config!(rules, config)
+        return unless filtering_present?(rules, config)
+
+        check_find_and_aggregate_present!(config)
+
+        check_find_and_aggregate_missing!(config)
+      end
+
+      def check_find_and_aggregate_missing!(config)
+        allowed_keys = Utility::Strings.format_string_array(ALLOWED_TOP_LEVEL_FILTER_KEYS, default: ' ')
+        keys_present = Utility::Strings.format_string_array(config.keys, default: ' ')
```

Comment:
Why is the special formatting necessary? Just doing a `.to_s` on a string array gets you the default behavior I think you're looking for:

```
irb(main):002:0> a = ['foo', 'bar']
=> ["foo", "bar"]
irb(main):003:0> s = a.to_s
=> "[\"foo\", \"bar\"]"
irb(main):004:0> puts s
["foo", "bar"]
=> nil
```

Reply:
+1 to that, seems like an unnecessary function that makes reading the code harder

Reply:
I thought it would be nice to have more control (omitting brackets for example, showing a space instead of an empty string -> not showing a space/default message could lead to "Some message:.", if no elements are present, which looks a bit weird IMO). But I see the point of just using the built-in `to_s` in favor of a cleaner codebase. I'll drop the class then 👍